### PR TITLE
Secret registry state machine

### LINKED ISSUE #1579: Take SecretRegistry into account in the state machine
Register filter for the Secret Registry

- [x] Registry a filter for all secrets registered on-chain
- [x] Create a state change for the learned secrets `ContractReceiveSecretReveal` - added in https://github.com/raiden-network/raiden/pull/1560
- [ ] Handle this state change with the proper transfer tasks
- [ ] add an additional mapping to the `NettingChannelEndState` e.g. `secrethashes_to_onchainunlockedlocks`, that should store the locks with secrets that have been registered on-chain; this mapping will keep the state clear, especially when the "remove expired locks" https://github.com/raiden-network/raiden/issues/193 functionality will be added (easier to differentiate between expired locks with or without on-chain registered secrets)

### PR DESCRIPTION
**Pull request by palango**
Fixes #1579

### CODE REVIEW

File: raiden/raiden_event_handler.py [hunk 1]
```diff
@@ -172,6 +173,13 @@ def handle_contract_send_secretreveal(
     secret_registry.register_secret(channel_close_event.secret)
 
 
+def handle_contract_receive_secretreveal(
+        raiden: 'RaidenService',
+        receive_secret_reveal_event: ContractReceiveSecretReveal,
+):
+    pass
```
Comment (palango) [inline raiden/raiden_event_handler.py]: What should be done here? Loredana told me that we shouldn't add the known secretes to some state. Anything else or do we just want to have them in the WAL?

File: raiden/blockchain/events.py [hunk 2]
```diff
@@ -47,12 +52,13 @@
 
 
 def get_contract_events(
-        chain,
+        chain: BlockChainService,
         abi,
```
Comment (hackaugusto) [inline raiden/blockchain/events.py]: this can be typed just as `Dict` for now

File: raiden/blockchain/events.py [hunk 3]
```diff
@@ -47,12 +52,13 @@
...
-        contract_address,
-        topics,
-        from_block,
-        to_block):
+        contract_address: Address,
+        topics: List[str],
+        from_block: BlockSpecification,
+        to_block: BlockSpecification,
+):
```
Comment (hackaugusto) [inline raiden/blockchain/events.py]: `-> List[Dict]`

File: raiden/blockchain/events.py [hunk 4]
```diff
@@ -77,14 +83,13 @@ def get_contract_events(
...
-        to_block='latest'):
-    """ Helper to get all events of the ChannelManagerContract at
-    `token_address`.
-    """
+        chain: BlockChainService,
+        channel_manager_address: Address,
+        events: List[str] = ALL_EVENTS,
+        from_block: BlockSpecification = 0,
+        to_block: BlockSpecification = 'latest',
+):
```
Comment (hackaugusto) [inline raiden/blockchain/events.py]: `-> List[Dict]`

File: raiden/blockchain/events.py [hunk 5]
```diff
@@ -97,14 +102,13 @@ def get_all_channel_manager_events(
...
-        to_block='latest'):
-    """ Helper to get all events of the Registry contract at
-    `registry_address`.
-    """
+        chain: BlockChainService,
+        registry_address: Address,
+        events: List[str] = ALL_EVENTS,
+        from_block: BlockSpecification = 0,
+        to_block: BlockSpecification = 'latest',
+):
```
Comment (hackaugusto) [inline raiden/blockchain/events.py]: `-> List[Dict]`

File: raiden/blockchain/events.py [hunk 6]
```diff
@@ -116,14 +120,13 @@ def get_all_registry_events(
...
-        to_block='latest'):
-    """ Helper to get all events of a NettingChannelContract at
-    `channel_identifier`.
-    """
+        chain: BlockChainService,
+        netting_channel_address: Address,
+        events: List[str] = ALL_EVENTS,
+        from_block: BlockSpecification = 0,
+        to_block: BlockSpecification = 'latest',
+):
```
Comment (hackaugusto) [inline raiden/blockchain/events.py]: -> List[Dict]

File: raiden/blockchain/events.py [hunk 7]
```diff
@@ -135,6 +138,27 @@ def get_all_netting_channel_events(
     )
 
 
+def get_all_secret_registry_events(
+        chain: BlockChainService,
+        secret_registry_address: Address,
+        events: List[str] = ALL_EVENTS,
+        from_block: BlockSpecification = 0,
+        to_block: BlockSpecification = 'latest',
+):
```
Comment (hackaugusto) [inline raiden/blockchain/events.py]: `-> List[Dict]`

File: raiden/transfer/mediated_transfer/mediator.py [hunk 8]
```diff
@@ -793,6 +821,7 @@ def secret_learned(
         secrethash,
         payee_address,
         new_payee_state,
+        from_onchain_secretreveal,
```
Comment (palango) [inline raiden/transfer/mediated_transfer/mediator.py]: Not the cleanest solution, but duplicating this function wouldn't be better either.

**Comment by palango** [comment]
@hackaugusto I added the `secrethashes_to_onchain_unlockedlocks` dict in the `NettingChannelEndState`. Please have a look.

File: raiden/blockchain/events.py [hunk 9]
```diff
@@ -135,6 +138,27 @@ def get_all_netting_channel_events(
...
+        secret_registry_address: Address,
+        events: List[str] = ALL_EVENTS,
+        from_block: BlockSpecification = 0,
+        to_block: BlockSpecification = 'latest',
+) -> List[Dict]:
+    """ Helper to get all events of a SecretRegistry. """
+
+    return get_contract_events(
+        chain,
+        raiden_contracts.contract_manager.CONTRACT_MANAGER.get_contract_abi(
```
Comment (hackaugusto) [inline raiden/blockchain/events.py]: why are we using a different contract manager to get the abi here?

Comment (pcppcp) [reply raiden/blockchain/events.py]: It's just temporary solution until the new contracts are integrated (to keep old & new contracts separated). When we switch over there'll be a single instance.

File: raiden/transfer/channel.py [hunk 10]
```diff
@@ -1220,14 +1255,29 @@ def register_secret_endstate(
...
 
 
+def register_onchain_secret_endstate(
+        end_state: NettingChannelEndState,
+        secret: typing.Secret,
+        secrethash: typing.SecretHash,
+) -> None:
+    if is_lock_locked(end_state, secrethash):
+        pendinglock = end_state.secrethashes_to_lockedlocks[secrethash]
+        del end_state.secrethashes_to_lockedlocks[secrethash]
```
Comment (loredanacirstea) [inline raiden/transfer/channel.py]: I think it is better to use `_del_lock` . At this point it is possible to have the `secret` in `secrethashes_to_unlockedlocks` in the case where the on-chain secret reveal transaction was sent by this node during the unsafe period.

File: raiden/transfer/mediated_transfer/mediator.py [hunk 11]
```diff
@@ -444,6 +445,33 @@ def set_secret(state, channelidentifiers_to_channels, secret, secrethash):
         )
 
 
+def set_onchain_secret(state, channelidentifiers_to_channels, secret, secrethash):
+    """ Set the secret to all mediated transfers.
+    It doesn't matter if the secret was learned through the blockchain or a
```
Comment (loredanacirstea) [inline raiden/transfer/mediated_transfer/mediator.py]: Code style - empty line after docstring title

File: raiden/transfer/channel.py (hunk 10 above)
Comment (palango) [reply raiden/transfer/channel.py]: Good point, but then it wouldn't be in `secrethashes_to_lockedlocks` any more, right?

Comment (loredanacirstea) [reply raiden/transfer/channel.py]: Correct. So it can either be in `secrethashes_to_lockedlocks` or `secrethashes_to_unlockedlocks`.

Comment (palango) [reply raiden/transfer/channel.py]: We missed that `is_lock_locked` only checks in  `secrethashes_to_lockedlocks`, so should be fine as it was before.

```py
secrethashes_to_lockedlocks
```

Comment (loredanacirstea) [reply raiden/transfer/channel.py]: My point was that locks can either be in `secrethashes_to_lockedlocks` , `secrethashes_to_unlockedlocks` or `secrethashes_to_onchain_unlockedlocks`. 

If you add it to `secrethashes_to_onchain_unlockedlocks` with `register_onchain_secret_endstate` , the lock could have been in either `secrethashes_to_lockedlocks` or `secrethashes_to_unlockedlocks` - we need to try to remove it from both. 

So `if is_lock_locked(end_state, secrethash):`  should be removed in my opinion.

File: raiden/transfer/channel.py [hunk 12]
```diff
@@ -1219,14 +1254,39 @@ def register_secret_endstate(
...
+        secret: typing.Secret,
+        secrethash: typing.SecretHash,
+) -> None:
+    # the lock might be in end_state.secrethashes_to_lockedlocks or
+    # end_state.secrethashes_to_unlockedlocks
+    # It should be removed from both and moved into secrethashes_to_onchain_unlockedlocks
+    pendinglock = None
+
+    if is_lock_locked(end_state, secrethash):
+        pendinglock = end_state.secrethashes_to_lockedlocks.get[secrethash]
```
Comment (loredanacirstea) [inline raiden/transfer/channel.py]: `get[secrethash]` -> `get(secrethash)`

Comment (palango) [reply raiden/transfer/channel.py]: thanks!

**Comment by LefterisJP** [comment]
Ready to go!

**Pull request merged**